### PR TITLE
Add refactoring to help number comments

### PR DESCRIPTION
**Pull request by jcouv**

### CODE REVIEW

**Comment by jcouv** [comment]
@mavasani This refactoring is rather narrow/specialized, as it helps with authoring compiler tests.
I wasn't sure in which project (or even repo) to put it. Is the location I picked appropriate?

I still need to make a couple of tweaks to this PR (fix some hard-coded strings), then I'll ping for review. Thanks

**Comment by mavasani** [comment]
@jcouv If this refactoring is only for Roslyn.sln, it should go into Roslyn.Diagnostics.Analyzers. However note that we do not have any refactorings in this repo yet. For that matter, this refactoring will not show up until https://github.com/dotnet/roslyn/pull/35685 is merged - Tagging @sharwell

**Comment by mavasani** [comment]
@jcouv If this refactoring is only for Roslyn.sln, it should go into Roslyn.Diagnostics.Analyzers. However note that we do not have any refactorings in this repo yet. For that matter, this refactoring will not show up until https://github.com/dotnet/roslyn/pull/35685 is merged - Tagging @sharwell

**Comment by jcouv** [comment]
Thanks. I moved the refactoring to `Roslyn.Diagnostics.Analyzers` and I'll wait for PR https://github.com/dotnet/roslyn/pull/35685

File: src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+{
+    /// <summary>
+    /// This refactoring looks for numbered comments `// N` or `// N, N+1, ...` on non-empty lines
+    /// and checks that the numbers are sequential. If they are not, the refactoring is offered.
+    ///
+    /// This pattern is commonly used by compiler tests.
+    /// Comments that don't look like numbered comments are left alone. For instance, any comment that contains alpha characters.
+    /// </summary>
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(NumberCommentslRefactoring)), Shared]
+    class NumberCommentslRefactoring : CodeRefactoringProvider
```
Comment (sharwell) [inline src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs]: 📝 `CodeRefactoringProvider` does not work via NuGet packages. This refactoring will be ignored by Visual Studio.

Comment (sharwell) [inline src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs]: 📝 `CodeRefactoringProvider` does not work via NuGet packages. This refactoring will be ignored by Visual Studio.

**Comment by mavasani** [comment]
@jcouv Are we still working on getting this through? If so, I would recommend that you just change the implementation to use a hidden diagnostic/code fixer to remove the dependency on https://github.com/dotnet/roslyn/pull/35685

**Comment by jcouv** [comment]
@mavasani I find switching to an analyzer/fixer trouble (typically takes me more time than I expect), so I was hoping the dependency issue would get resolved. Do you  know what is holding it up?

**Comment by jcouv** [comment]
I see that https://github.com/dotnet/roslyn/pull/35685 has been merged (thanks @sharwell!), so I think we can move forward with this PR.

**Comment by jcouv** [comment]
@mavasani for review. Thanks

File: src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs (hunk 1 above)
Comment (mavasani) [inline src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs]:
```suggestion
    internal sealed class NumberCommentslRefactoring : CodeRefactoringProvider
```

File: src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span);
+
+            if (node is LiteralExpressionSyntax literal &&
+                literal.Kind() == SyntaxKind.StringLiteralExpression &&
+                !IsProperlyNumbered(literal.Token.ValueText))
+            {
+                var action = CodeAction.Create("TODO", c => FixCommentsAsync(context.Document, literal, c));
```
Comment (mavasani) [inline src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs]: Code action title must be a localized string. Also, is "TODO" an informative enough title here or you meant to replace this with a better title?

File: src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Test.Utilities;
+using Xunit;
+
+namespace MetaCompilation.Analyzers.UnitTests
+{
+    public abstract class CodeRefactoringVerifier : DiagnosticVerifier
```
Comment (mavasani) [inline src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs]: @sharwell Does the new testing library have APIs for code refactoring testing?

File: src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Test.Utilities;
+using Xunit;
+
+namespace MetaCompilation.Analyzers.UnitTests
```
Comment (mavasani) [inline src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs]: Should this file be moved under https://github.com/dotnet/roslyn-analyzers/tree/master/src/Test.Utilities? Namespace should also be changed accordingly.

File: src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs (hunk 3 above)
Comment (sharwell) [reply src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs]: Not yet.

File: src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs (hunk 2 above)
Comment (jcouv) [reply src/Roslyn.Diagnostics.Analyzers/Core/NumberCommentslRefactoring.cs]: Oops, sorry about that. TODO is definitely meant to be fixed before merge. I'll fix today

---
In reply to: [364427115](https://github.com/dotnet/roslyn-analyzers/pull/2588#discussion_r364427115) [](ancestors = 364427115)

File: src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs (hunk 4 above)
Comment (jcouv) [reply src/MetaCompilation.Analyzers/UnitTests/Verifiers/CodeRefactoringVerifier.cs]: The trouble is that this type derives from `MetaCompilation.Analyzers.UnitTests.DiagnosticVerifier`. Should we move `DiagnosticVerifier` as well?

---
In reply to: [364432464](https://github.com/dotnet/roslyn-analyzers/pull/2588#discussion_r364432464) [](ancestors = 364432464)

**Comment by jcouv** [comment]
Some manual validation:
image

Strange... the setup seemed to work for this refactoring. I'm not sure why it didn't seem to work for my other PR (related to nullability in shipped API files, https://github.com/dotnet/roslyn-analyzers/pull/3125)

**Pull request merged**